File: examples/go/webhook_handler.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/yourdudeken/mpesa-sdk/client"
	"github.com/yourdudeken/mpesa-sdk/types"
	"github.com/yourdudeken/mpesa-sdk/webhooks"
)

func main() {
	mpesa := client.NewClient(types.MpesaConfig{
		ConsumerKey:    os.Getenv("MPESA_CONSUMER_KEY"),
		ConsumerSecret: os.Getenv("MPESA_CONSUMER_SECRET"),
		Environment:    types.Sandbox,
		Passkey:        os.Getenv("MPESA_PASSKEY"),
	})

	wh := webhooks.NewManager()

	wh.On(webhooks.EventSTKCallback, func(et webhooks.EventType, payload interface{}) {
		if result, ok := payload.(types.STKCallbackResult); ok {
			if result.Success {
				receipt := "<nil>"
				if result.ReceiptNumber != nil {
					receipt = *result.ReceiptNumber
				}
				amount := 0.0
				if result.Amount != nil {
					amount = *result.Amount
				}
				fmt.Printf("[STK] Payment: %s KES %.0f\n", receipt, amount)
			} else {
				fmt.Printf("[STK] Failed: %s (code: %d)\n", result.ResultDescription, result.ResultCode)
			}
		}
	})

	wh.On(webhooks.EventB2CResult, func(et webhooks.EventType, payload interface{}) {
		fmt.Println("[B2C] Result received")
	})

	wh.On(webhooks.EventAccountBalance, func(et webhooks.EventType, payload interface{}) {
		fmt.Println("[Balance] Account balance data received")
	})

	wh.On(webhooks.EventTransactionStatus, func(et webhooks.EventType, payload interface{}) {
		fmt.Println("[Status] Transaction status data received")
	})

	wh.On(webhooks.EventReversalResult, func(et webhooks.EventType, payload interface{}) {
		fmt.Println("[Reversal] Reversal result received")
	})

	wh.On(webhooks.EventC2BValidation, func(et webhooks.EventType, payload interface{}) {
		fmt.Println("[C2B] Validation request received")
	})

	ctx := context.Background()
	resp, err := mpesa.STKPush(ctx, types.STKPushRequest{
		BusinessShortCode: 174379,
		TransactionType:   types.CustomerPayBillOnline,
		Amount:            5000,
		PartyA:            254722000000,
		PartyB:            174379,
		PhoneNumber:       254722000000,
		CallBackURL:       "https://api.yourdomain.com/mpesa/callback",
		AccountReference:  "ORDER-12345",
		TransactionDesc:   "Order payment",
	})
	if err != nil {
		log.Fatalf("STK Push failed: %v", err)
	}
	fmt.Printf("Order initiated: %s\n", resp.CheckoutRequestID)

	payload := types.STKCallbackPayload{}
	payload.Body.StkCallback.MerchantRequestID = resp.MerchantRequestID
	payload.Body.StkCallback.CheckoutRequestID = resp.CheckoutRequestID
	payload.Body.StkCallback.ResultCode = 0
	payload.Body.StkCallback.ResultDesc = "Success"
	payload.Body.StkCallback.CallbackMetadata = &struct {
		Item []struct {
			Name  string      `json:"Name"`
			Value interface{} `json:"Value"`
		} `json:"Item"`
	}{
		Item: []struct {
			Name  string      `json:"Name"`
			Value interface{} `json:"Value"`
		}{
			{Name: "Amount", Value: float64(5000)},
			{Name: "MpesaReceiptNumber", Value: "NLA12345XX"},
		},
	}

	// Parse and dispatch through the webhook manager
	result := client.ParseSTKCallback(payload)
	wh.Emit(webhooks.EventSTKCallback, result)

	if result.Success {
		receipt := "<nil>"
		if result.ReceiptNumber != nil {
			receipt = *result.ReceiptNumber
		}
		amount := 0.0
		if result.Amount != nil {
			amount = *result.Amount
		}
		fmt.Printf("Payment confirmed: %s KES %.0f\n", receipt, amount)
	}
}

File: examples/python/webhook-handler.py
import os
import time
from mpesa import Mpesa, WebhookManager

client = Mpesa({
    "consumer_key": os.environ["MPESA_CONSUMER_KEY"],
    "consumer_secret": os.environ["MPESA_CONSUMER_SECRET"],
    "environment": "sandbox",
    "passkey": os.environ.get("MPESA_PASSKEY"),
    "initiator_name": os.environ.get("MPESA_INITIATOR_NAME"),
    "security_credential": os.environ.get("MPESA_SECURITY_CREDENTIAL"),
})

webhooks = WebhookManager()


@webhooks.on("stk:callback")
def handle_stk(event_type, payload):
    result = webhooks.parse_stk_callback(payload)
    if result["success"]:
        print(f"[STK] Payment: {result['receipt_number']} KES {result['amount']}")
    else:
        print(f"[STK] Failed: {result['result_description']}")


@webhooks.on("b2c:result")
def handle_b2c(event_type, payload):
    print(f"[B2C] Result received")


@webhooks.on("account:balance")
def handle_balance(event_type, payload):
    print(f"[Balance] Account balance data received")


@webhooks.on("transaction:status")
def handle_status(event_type, payload):
    print(f"[Status] Transaction status data received")


@webhooks.on("reversal:result")
def handle_reversal(event_type, payload):
    print(f"[Reversal] Reversal result received")


@webhooks.on("c2b:validation")
def handle_c2b_validation(event_type, payload):
    trans_id = payload.get("TransID", "<missing>")
    print(f"[C2B] Validation: {trans_id}")
    return webhooks.parse_c2b_validation_response(accept=True)


SHORTCODE = int(os.environ.get("MPESA_SHORTCODE", "174379"))


def production_workflow():
    # 1. Initiate STK Push
    stk = client.stk_push({
        "BusinessShortCode": SHORTCODE,
        "TransactionType": "CustomerPayBillOnline",
        "Amount": 5000,
        "PartyA": 254722000000,
        "PartyB": SHORTCODE,
        "PhoneNumber": 254722000000,
        "CallBackURL": "https://api.yourdomain.com/mpesa/callback",
        "AccountReference": "ORDER-12345",
        "TransactionDesc": "Order payment",
    })
    print(f"1. STK Push sent: {stk.CheckoutRequestID}")

    # 2. Query status after delay
    time.sleep(5)
    status = client.stk_query({
        "BusinessShortCode": str(SHORTCODE),
        "CheckoutRequestID": stk.CheckoutRequestID,
    })
    print(f"2. Payment status: {status.ResultDesc}")

    # 3. Check account balance
    balance = client.account_balance({
        "Initiator": os.environ["MPESA_INITIATOR_NAME"],
        "SecurityCredential": os.environ["MPESA_SECURITY_CREDENTIAL"],
        "CommandID": "AccountBalance",
        "PartyA": SHORTCODE,
        "IdentifierType": 4,
        "Remarks": "Reconcile balance",
        "QueueTimeOutURL": "https://api.yourdomain.com/mpesa/queue",
        "ResultURL": "https://api.yourdomain.com/mpesa/balance-result",
    })
    print(f"3. Balance query sent: {balance.OriginatorConversationID}")

    # 4. Simulate receiving a webhook
    webhooks.emit("stk:callback", {
        "Body": {
            "stkCallback": {
                "MerchantRequestID": stk.MerchantRequestID,
                "CheckoutRequestID": stk.CheckoutRequestID,
                "ResultCode": 0,
                "ResultDesc": "Success",
                "CallbackMetadata": {
                    "Item": [
                        {"Name": "Amount", "Value": 5000},
                        {"Name": "MpesaReceiptNumber", "Value": "NLA12345XX"},
                    ],
                },
            },
        },
    })


if __name__ == "__main__":
    production_workflow()
